.github/workflows/ci.yml

```diff
@@ -0,0 +1,352 @@
+
+      - name: Install rust toolchain
+        working-directory: ${{ env.RUST_WORKSPACE }}
+        run: rustup show
+
+      - uses: Swatinem/rust-cache@d12701459954fec471b2d34cdf7ea3374b026383 # v1
+        with:
+          working-directory: ${{ env.RUST_WORKSPACE }}
+
+      - name: cargo clippy
+        working-directory: ${{ env.RUST_WORKSPACE }}
+        run: |
+          # workaround for some code being created after rust wants to parse it
+          cargo check --quiet 2>/dev/null || :
+          cargo clippy --all-targets -- --deny warnings
+
+  cargo-test:
+    needs: selection
+    if: ${{ needs.changes.outputs.rust == 'true' }}
+    runs-on: ubuntu-20.04
+    timeout-minutes: 20
+    steps:
+      - uses: actions/checkout@ec3a7ce113134d7a93b817d10a8272cb61118579 # v2.4.0
+
+      - name: Install rust toolchain
+        working-directory: ${{ env.RUST_WORKSPACE }}
+        run: rustup show
+
+      - uses: Swatinem/rust-cache@d12701459954fec471b2d34cdf7ea3374b026383 # v1
+        with:
+          working-directory: ${{ env.RUST_WORKSPACE }}
+
+      - name: Run tests
+        working-directory: ${{ env.RUST_WORKSPACE }}
+        run: |
+          # workaround for some code being created after rust wants to parse it
+          cargo check --quiet 2>/dev/null || :
+          cargo test --all-targets --quiet
+          cargo test --doc --quiet
+
+  cargo-doc:
+    needs: selection
+    if: ${{ needs.changes.outputs.rust == 'true' }}
+    runs-on: ubuntu-20.04
+    timeout-minutes: 20
+    steps:
+      - uses: actions/checkout@ec3a7ce113134d7a93b817d10a8272cb61118579 # v2.4.0
+
+      - name: Install rust toolchain
+        working-directory: ${{ env.RUST_WORKSPACE }}
+        run: rustup show
+
+      - uses: Swatinem/rust-cache@d12701459954fec471b2d34cdf7ea3374b026383 # v1
+
+      - name: cargo doc
+        working-directory: ${{ env.RUST_WORKSPACE }}
+        run: |
+          # workaround for some code being created after rust wants to parse it
+          cargo check --quiet 2>/dev/null || :
+          cargo doc --all-features --no-deps --document-private-items
+
+    # this is an helper that needs all the real leafs of the workflow.
+    # It makes easier notify_main_failure because we only need to check
+    # for this job
+
+  cargo-ok:
+    name: cargo-ok
+    needs:
+      - cargo-format
+      - cargo-clippy
+      - cargo-test
+      - cargo-doc
+    runs-on: ubuntu-20.04
+    steps:
+      - run: exit 0
+  
+  ci-ok:
+    name: ci-ok
+    needs: [selection, dart-doc, cargo-ok]
+    if: ${{always()}}
+    runs-on: ubuntu-20.04
+    timeout-minutes: 5
+    steps:
+      - name: CI OK Testing
+        if: ${{ needs.selection.outputs.dart == 'true' }} || ${{ needs.selection.outputs.rust == 'true' }}
+        run: |
+          if ( ${{ needs.selection.outputs.dart == 'true' }} && ${{ needs.selection.outputs.rust == 'true' }} && ${{ needs.dart-doc.result == 'success'}} && ${{ needs.cargo-ok.result == 'success'}} )
```

Comment:
Can't we write all of this as:
`(outputs.dart implies dart.result == success) and (outputs.rust implies rust.result == success)`
?
instead of having multiple cases?

Reply:
Ok, I will try.

Reply:
I made some changes, let me know if that was what you were referring to.